src/disk_cache.rs

```diff
@@ -0,0 +1,955 @@
+}
+
+struct DiskCacheEntry {
+    root_folder: std::path::PathBuf,
+    location: object_store::path::Path,
+    part_size: usize,
+}
+
+#[allow(unused)]
+type PartID = usize;
+
+#[allow(unused)]
+impl DiskCacheEntry {
+    /// save the GetResult to the disk cache, a GetResult may be transformed into multiple part
+    /// files and a meta file. please note that the `range` in the GetResult is expected to be
+    /// aligned with the part size.
+    pub async fn save_result(&self, result: GetResult) -> object_store::Result<usize> {
+        assert!(result.range.start % self.part_size == 0);
+        assert!(result.range.end % self.part_size == 0 || result.range.end == result.meta.size);
+
+        self.save_meta(&result.meta).await?;
+
+        let mut buffer = BytesMut::new();
+        let mut part_number = result.range.start / self.part_size;
+        let object_size = result.meta.size;
+
+        let mut stream = result.into_stream();
+        while let Some(chunk) = stream.next().await {
+            let chunk = chunk?;
+            buffer.extend_from_slice(&chunk);
+
+            while buffer.len() >= self.part_size {
+                let to_write = buffer.split_to(self.part_size);
+                self.save_part(part_number, to_write.as_ref()).await?;
+                part_number += 1;
+            }
+        }
+
+        // if the last part is not fully filled, save it as the last part.
+        if !buffer.is_empty() {
+            self.save_part(part_number, buffer.as_ref()).await?;
+            return Ok(object_size);
+        }
+
+        Ok(object_size)
+    }
+
+    pub async fn read_part(&self, part_id: PartID) -> object_store::Result<Option<Bytes>> {
+        let part_path = self.make_part_path(part_id, false);
+
+        // if the part file does not exist, return None
+        let exists = fs::try_exists(&part_path).await.unwrap_or(false);
+        if !exists {
+            return Ok(None);
+        }
+
+        let file = File::open(&part_path).await.map_err(wrap_io_err)?;
+        let mut reader = tokio::io::BufReader::new(file);
+        let mut buffer = Vec::new();
+        reader.read_to_end(&mut buffer).await.map_err(wrap_io_err)?;
+        Ok(Some(Bytes::from(buffer)))
+    }
+
+    pub async fn read_meta(&self) -> object_store::Result<Option<ObjectMeta>> {
+        let meta_path = self.make_meta_path();
+
+        let exists = fs::try_exists(&meta_path).await.unwrap_or(false);
+        if !exists {
+            return Ok(None);
+        }
+
+        // TODO: process not found err here instead of check exists
+        let file = File::open(&meta_path).await.map_err(wrap_io_err)?;
+        let mut reader = tokio::io::BufReader::new(file);
+        let mut content = String::new();
+        reader
+            .read_to_string(&mut content)
+            .await
+            .map_err(wrap_io_err)?;
+
+        let meta: DiskCacheMeta = serde_json::from_str(&content).map_err(wrap_io_err)?;
+
+        Ok(Some(ObjectMeta {
```

Comment:
i think a checksum is needed finally. however having a checksum needs some design about the data layout in the files, i think it can be seperated as a standalone feature? we can easily migrate to the new cache file format with a simple file name versioning.

this PR attempts to make every filesystem write atomic. I think the cached data will not easily become inconsistent unless the filesystem corrupts the data (i belive it actually happens).

Reply:
+1 for checksum, and +1 for keeping it as a separate GH issue. @flaneur2020 mind opening it so we don't lose track?